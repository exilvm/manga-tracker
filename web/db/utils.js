const format = require('pg-format');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === '22P02') {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
