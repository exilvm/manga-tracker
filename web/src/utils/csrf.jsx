import { createContext, useContext } from 'react';

const CSRFContext = createContext(undefined);
export const CSRFProvider = CSRFContext.Provider;

export const useCSRF = () => useContext(CSRFContext);

/**
 * Returns the useCSRF token in the props property with the key _csrf
 * @param req
 * @returns {{props: {_csrf: String}}}
 */
export const csrfProps = ({ req }) => ({
  props: { _csrf: req.csrfToken() },
});

export const csrfHeader = (csrf) => ({
  'X-CSRF-Token': csrf,
});
