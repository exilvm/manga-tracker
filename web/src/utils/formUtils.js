export const showErrorAlways = ({
  meta: { submitError, dirtySinceLastSubmit, error },
}) => !!(((submitError && !dirtySinceLastSubmit) || error));
