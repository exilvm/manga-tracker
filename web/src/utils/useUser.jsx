import { createContext, useContext, useMemo } from 'react';


const UserContext = createContext(undefined);
export const UserProvider = UserContext.Provider;

export const useUser = () => {
  const user = useContext(UserContext);
  return useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isAdmin: user && user.admin,
  }), [user]);
};
