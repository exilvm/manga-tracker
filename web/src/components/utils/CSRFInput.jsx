import { Field } from 'react-final-form';
import { useCSRF } from '../../utils/csrf';

export default function CSRFInput() {
  const csrf = useCSRF();
  return (
    <Field
      name='_csrf'
      component='input'
      type='hidden'
      defaultValue={csrf}
    />
  );
}
