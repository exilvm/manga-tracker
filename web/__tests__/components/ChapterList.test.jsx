import React from 'react';
import { create } from 'react-test-renderer';
import { mockUTCDates } from '../utils';
import ChapterList from '../../src/components/ChapterList';

describe('Chapter list should render correctly', () => {
  mockUTCDates();
  const chapters = [
    {
      title: 'Z=157: Same Time, Same Place',
      chapter_number: 157,
      release_date: new Date(1593964800000),
      group: 'Shueisha',
      service_id: 1,
      chapter_url: 'https://mangaplus.shueisha.co.jp/titles/1007322',
    },
    {
      title: 'Z=156: Two Scientists',
      chapter_number: 156,
      release_date: new Date(null),
      group: 'MangaPlus',
      service_id: 2,
      chapter_url: 'https://mangadex.org/title/938629',
    },
    {
      title: 'Z=156: Two Scientists',
      chapter_number: 156,
      release_date: new Date(1593187200000),
      group: 'Shueisha',
      service_id: 1,
      chapter_url: 'https://mangaplus.shueisha.co.jp/titles/1007024',
    },
  ];

  it('with chapters', () => {
    const tree = create(
      <ChapterList chapters={chapters} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('without chapters', () => {
    const tree = create(
      <ChapterList chapters={[]} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('with chapters and edit', () => {
    const tree = create(
      <ChapterList chapters={chapters} editable />
    );

    expect(tree).toMatchSnapshot();
  });
});
