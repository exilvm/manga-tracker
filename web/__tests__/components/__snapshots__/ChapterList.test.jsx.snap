// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chapter list should render correctly with chapters 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root"
    role="table"
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role="row"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Title
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Chapter
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Released
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Group
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role="rowgroup"
    >
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=157: Same Time, Same Place
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          157
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Jul 5th 2020, 16:00
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Shueisha
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=156: Two Scientists
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          156
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Unknown
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          MangaPlus
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=156: Two Scientists
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          156
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Jun 26th 2020, 16:00
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Shueisha
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Chapter list should render correctly with chapters and edit 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root"
    role="table"
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role="row"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": undefined,
            }
          }
        />
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Title
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Chapter
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Released
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Group
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": undefined,
            }
          }
        />
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role="rowgroup"
    >
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <div
            className="makeStyles-editCell-1"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              name="edit"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=157: Same Time, Same Place
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          157
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Jul 5th 2020, 16:00
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Shueisha
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            name="delete"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <div
            className="makeStyles-editCell-1"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              name="edit"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=156: Two Scientists
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          156
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Unknown
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          MangaPlus
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            name="delete"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role="row"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <div
            className="makeStyles-editCell-1"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              name="edit"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            style={
              Object {
                "textDecoration": "none",
              }
            }
            target="_blank"
          >
            <span>
              Z=156: Two Scientists
            </span>
          </a>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          156
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Jun 26th 2020, 16:00
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
        >
          Shueisha
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
          role="cell"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            name="delete"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Chapter list should render correctly without chapters 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root"
    role="table"
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role="row"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Title
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Chapter
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Released
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          colSpan={1}
          onClick={[Function]}
          role="columnheader"
          scope="col"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          title="Toggle SortBy"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Group
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role="rowgroup"
    />
  </table>
</div>
`;
